### PR TITLE
externalconn: support `SHOW CREATE EXTERNAL CONNECTION`

### PR DESCRIPTION
**Pull request by adityamaru**
This change adds support for `SHOW CREATE EXTERNAL CONNECTION`
and `SHOW CREATE ALL EXTERNAL CONNECTIONS` that displays the
connection name and the statement used to create the external
connection. This displays unredacted information of the underlying
resource and is therefore restricted to admin only or by the
owner of the external connection object.

Note, synthetic privileges do not have a concept of external
connections at the moment. So, the operations are limited to
admin only until that functionality is added.

Informs: #85905

Release note (sql change): Add support for `SHOW CREATE EXTERNAL CONNECTION`
and `SHOW CREATE ALL EXTERNAL CONNECTIONS` that displays the
connection name and the unredacted query used to create the external
connection. This can only be run by users of the admin role today.

Release justification: low risk change to improve observability into external connections

### CODE REVIEW

**Comment by benbardin** [comment]
I know we have this in other places, but why do we want this? "Show me the SQL command used to create the thing" seems less helpful than "show me the helpful information about the thing directly"

**Comment by adityamaru** [comment]
> Show me the SQL command used to create the thing

If a user created an external connection foo with lengthy base64 encoded credentials and several still-to-be-added configuration options, and then they want to create a copy of this external connection just with a slightly different bucket or option or name they can copy paste the SQL from this command to recreate the external connection. This was pretty much the reason we added `SHOW CREATE SCHEDULE` as well. 

> show me the helpful information about the thing directly

As for this, we will have this too in `SHOW EXTERNAL CONNECTION foo` for which I need to resurrect https://github.com/cockroachdb/cockroach/pull/85913. That output will redact all secrets and will not reveal secrets that would enable one to recreate an external connection.

File: pkg/cloud/externalconn/connection.go [hunk 1]
```diff
@@ -21,6 +21,15 @@ import (
 // an External Connection object. This interface should expose read-only
 // methods that are required to interact with the External Connection object.
 type ExternalConnection interface {
+	// CreateConnectionStatement returns a `CREATE EXTERNAL CONNECTION` statement
+	// that is functionally equivalent to the statement that created the external
+	// connection in the first place.
+	//
+	// NB: The returned string will contain unredacted secrets and should not be
+	// persisted.
+	CreateConnectionStatement() string
```
Comment (benbardin) [inline pkg/cloud/externalconn/connection.go:30]: nit: maybe build "unredacted" into the name? `UnredactedConnectionStatement() string`? Also feels a bit weird to call this a "Create" when it returns a string

File: pkg/cloud/externalconn/connectionpb/connection.go [hunk 2]
```diff
@@ -22,6 +22,17 @@ func (d *ConnectionDetails) Type() ConnectionType {
...
 		return TypeStorage
 	default:
-		panic(errors.AssertionFailedf("ConnectionDetails.Type called on a details with an unknown type: %T", d.Provider.String()))
+		panic(errors.AssertionFailedf("ConnectionDetails.Type called on a details with an unknown type: %s", d.Provider.String()))
+	}
+}
+
+// UnredactedURI returns the unredacted URI of the resource represented by the
+// External Connection.
+func (d *ConnectionDetails) UnredactedURI() string {
```
Comment (benbardin) [inline pkg/cloud/externalconn/connectionpb/connection.go:31]: Yeah, see? Like this! This is great :)

**Review by benbardin — APPROVED** [review]

File: pkg/cloud/externalconn/connection.go (hunk 1 above)
Comment (adityamaru) [reply pkg/cloud/externalconn/connection.go:30]: good point, changed.

**Comment by adityamaru** [comment]
TFTR!

bors r=benbardin

**Comment by adityamaru** [comment]
bors r=benbardin

**Comment by adityamaru** [comment]
failure is `TestMergeQueueWithSlowNonVoterSnaps` which is unrelated

bors r=benbardin

**Pull request merged**